Faster changing gather peers

noaonoszko opened the pull request:
I have tested the validator and miner but not the aggregation server.

joellidin commented on src/tplr/comms.py:
```diff
@@ -1450,6 +1493,82 @@ async def post_start_window(self, start_window: int):
 
+    async def get_peer_list(self) -> tuple[int, int] | None:
+        tplr.logger.info("Starting to look for a peer list on a validator bucket")
+        while True:
```
I got stuck in this loop while testing. It seems like the validator never uploaded the first one.

joellidin commented on neurons/validator.py:
```diff
@@ -382,18 +370,54 @@ async def run(self):
+                    f"Time to create and post a new peer list because {reason}"
+                )
+                if self.last_peer_update_window is None:
+                    success = self.select_initial_peers()
```
Should probably check if this was a success or not.

joellidin commented on src/tplr/comms.py:
```diff
@@ -1450,6 +1493,82 @@ async def post_start_window(self, start_window: int):
+                if len(keys) == 0:
+                    tplr.logger.info("No peer list file found on bucket")
+                    return None
+                pattern = rf"^{PEERS_FILE_PREFIX}(?P<window>\d+)_v{__version__}\.json$"
```
Should filter keys from this patterna as well and not only `peers_` since otherwise everyone gets into a loop.

noaonoszko replied:
Damn good catch

joellidin commented on neurons/validator.py:
```diff
@@ -1626,6 +1649,157 @@ async def run(self):
+            )
+            tplr.logger.info(f"Not time to replace peers yet: {reason}")
+
+    def select_initial_peers(self) -> bool:
```
So a problem right now. For this to work we are checking active peers. But the miner will not start the training and become active. An alternative would be to not check for active peers as a initial selection?

joellidin replied:
Then the first thing the highest staked validator could do is post a start_window with the initial peers. Allowing everyone to start of from the same point? Even though the success rate might not be the best it would rotate quickly for the next peers? Wdyt @distributedstatemachine @noaonoszko?

noaonoszko replied:
You're right, training won't start because the miner continues if it fails to gather:
```python
            if gather_result is None:
                tplr.logger.error(
                    "Failed to gather gradients from peers. Waiting for next window."
                )
                while self.current_window == step_window:
                    await asyncio.sleep(0.1)
                continue
```
The plan was to accept that there wouldn't be any gathers for the first three or so windows, but I forgot to handle this in the code. Not using active peers initially could be ok but presents two potential problems:
1. If only one of the selected peers are active, I'm afraid the loss will increase. Can't explain it theoretically but I've observed it locally.
2. We need to filter out validators. Guess possible though, maybe based on stake?

noaonoszko replied:
Actually no, training does start since the continue is way after. If it doesn't there's some other problem, hopefully fixed by my fixup commits that I'll push later.

joellidin replied:
Validator should not have any incentives right? I was trying it out with just moving the initial gather peers  (with some modifications to the method) before the while loop and then update it as soon it receives active peers.

joellidin replied:
Needs more testing but seemed to work for the start of a run.

noaonoszko replied:
I'm confident there's a way to exclude validators so rn my only concern with not basing initial selection on active peers is point 1. (Although filtering on incentives doesn't work because there might nog be enough miners with incentive.)

I don't have anything more to say for now so let me know what the next step is regarding this.

joellidin commented on neurons/aggregator.py:
```diff
@@ -453,6 +453,56 @@ def handler(event):
+        # Get next peers
+        if (
+            self.next_peers is None  # next peers are not fetched yet
+            and self.peers_update_window  # they should be on bucket by now
```
This does not exist either.

joellidin commented on neurons/miner.py:
```diff
@@ -748,6 +735,56 @@ async def run(self):
             while self.current_window == step_window:
                 await asyncio.sleep(0.1)
 
+    async def update_peers(self, step_window: int) -> None:
```
Maybe move this logic to something like `src/tplr/neurons`?

noaonoszko replied:
Yes it's super ugly that this method exists almost identical in three classes (`Validator`, `Miner` and `AggregationServer`). But there are so many if/else cases where several attributes are defined so it feels almost impossible. I think the best solution would be to have a `Neuron` super class that implements this method. Wdyt?

joellidin replied:
I think we could do something similar like we did for the catchup where we just pass the instance. All the required class variables is the same for all three right? We could refactor this later to use a super class or a `Protocol` to make it cleaner like you suggested. I am working a bit now to see if I could get this working.

noaonoszko replied:
Good idea! I'll pass the instance.

joellidin commented on neurons/aggregator.py:
Maybe don't update active peers in this one if it should come from the validator. Or maybe it is the aggregator that should decide which peers to use?

joellidin commented on neurons/validator.py:
```diff
@@ -382,18 +370,54 @@ async def run(self):
+
             self.comms.update_peers_with_buckets()
-            self.peers = self.comms.peers
+            self.peer_start = tplr.T()
```
Here as well.

joellidin commented on neurons/validator.py:
```diff
@@ -382,18 +370,54 @@ async def run(self):
                 f"Processing window: {self.sync_window} current: {self.current_window}"
             )
 
-            peer_start = tplr.T()
```
This is a breaking change. Maybe revert back to `peer_start = tplr.T()`

joellidin reviewed (changes requested):
As it was right now, it did not run in the test setup. There are some issues I pointed out here. But when I hotfixed some of these it seems to be running. But in the beginning the aggregator and miners don't gather at all from the initial gather peers. I suggest that these are initialized in the `start_window.json` or something similar instead.

noaonoszko commented:
Looks like I messed up everything when I updated the branch 😫 Will fix tomorrow (Sunday)

noaonoszko commented:
Not ready for review yet, will test on local chain first.

In addition to fixing your comments, I fixed a bug: I changed so that the selected peers from `Validator.select_initial_peers` and `Validator.select_next_peers` are returned instead of setting `self.comms.peers`, because otherwise the validator will use the new list before everybody else.

noaonoszko commented:
Ready for rewiew.

Validator+2 miners tested on local chain: It runs, loss decreases, peer lists are created as expected. Aggregation server not tested.

Running a little longer now with 3 miners just to be sure.

I think something should be done about the name clash between [neurons/](https://github.com/tplr-ai/templar/tree/main/neurons) and [src/tplr/neurons.py](https://github.com/tplr-ai/templar/blob/main/src/tplr/neurons.py). I had to do this because of it to make my tests run: ddf8162.

joellidin commented:
> I just force-pushed to update the branch. No merge conflicts so _should_ be safe.
> 
> Wdyt @joellidin about making `peers` an attribute of the `Validator`, `Miner` and `AggregationServer`? I see no point for it to be a `Comms` attribute. I made an arbitrary decision to make it a `Comms` attribute when I started working on this, as I noticed that `peers` was an attribute in both 😱 I don't want to make this PR harder to review so I can do it in a separate PR also.
> 
> In fact, and as an aside, I don't see a reason for these "neurons" to have a `Comms` instance as an attribute at all as `Comms` doesn't change the neurons' state. It just causes confusion about whether/how they're coupled.

I think this makes a lot of sense. I think it makes sense to have a `Comms` instance available so we can access the communcation. At least as it is for the current state of the code. Or how would you suggest us to change this? I think it could be a separate PR to improve this you suggested so we can try and get this in as it is. But very valid point and would be a great change.

joellidin commented:
I will be trying this out with the aggregator later today. One thing I noticed which I did not write in my last review is that the validator does not update the gather peers when a gather peer get inactive. I don't know if you have fixed this, but I think we should fix that as well before merging. I will come back later after I have tested it.

joellidin commented:
> I will be trying this out with the aggregator later today. One thing I noticed which I did not write in my last review is that the validator does not update the gather peers when a gather peer get inactive. I don't know if you have fixed this, but I think we should fix that as well before merging. I will come back later after I have tested it.

This might mostly be noticeable in small scale testing though. When the validator changes the weight more often for many peers it might not happen so often. But it could. Also, what is the expected behavior if a gather peer get 0 weight but we don't have enough candidates to replace it. Should that be dropped as well as long as we are above the minimum gather peers target?

noaonoszko commented:
I realize it's not as straight-forward as I first thought. I was thinking simply defining the `Comms` instance outside of the neuron class but in the same file, but the `Comms` constructor actually takes some attributes of the neuron right now. But it doesn't have to:
- `config`: Is a classmethod of the neuron.
- `wallet`: Comes from the config. (And is only used to create the `tmp` directory using the hotkey ss58 address.)
- `uid`: Is created using the `wallet`.
- `totalks`: Not used at all in `Comms`.

This is off-topic but I might open a PR to separate these later.

noaonoszko commented:
> the validator does not update the gather peers when a gather peer get inactive.
Good catch, I'll fix.

> Also, what is the expected behavior if a gather peer get 0 weight but we don't have enough candidates to replace it. Should that be dropped as well as long as we are above the minimum gather peers target?
I was thinking we keep it but I'm open to dropping it, maybe that's better?

joellidin commented on neurons/aggregator.py:
```diff
@@ -242,7 +246,10 @@ async def process_window(self):
             )
 
         # Use comms to select gather peers
-        self.comms.set_gather_peers()
```
We have a reference of this still on line 448 in the aggregator

joellidin commented on src/tplr/comms.py:
```diff
@@ -51,6 +52,10 @@
+PEERS_FILE_PREFIX = "peers_"
+
+# Types
+PeerArray = np.ndarray[Any, np.dtype[np.int64]]
```
Why can't we use the normal list? I don't see the reason to use a numpy array for this

noaonoszko replied:
I wanted it to be consistent with `metagraph.uids`, which is an array of ints.

joellidin replied:
Let's keep it like that for now.

joellidin commented on neurons/validator.py:
```diff
@@ -1626,6 +1656,105 @@ async def run(self):
+        try:
+            tplr.logger.info("Starting selection of initial gather peers")
+            # 1. Add top incentive peers
+            uid_to_non_zero_incentive = {
+                uid: incentive
+                for uid, incentive in zip(self.metagraph.uids, self.metagraph.I)
+                if incentive > 0 and uid in self.comms.active_peers
+            }
+            top_incentive_peers = sorted(
+                uid_to_non_zero_incentive,
+                key=uid_to_non_zero_incentive.get,
+                reverse=True,
+            )[: self.hparams.max_topk_peers]
+
+            # Convert to list to ensure it's not a dict_keys object
+            top_incentive_peers = np.array(top_incentive_peers, dtype=np.int64)
```
```suggestion
            uid_to_non_zero_incentive = {
                uid: incentive
                for uid, incentive in zip(self.metagraph.uids, self.metagraph.I)
                if incentive > 0 and uid in self.comms.active_peers
            }
            top_incentive_peers = sorted(
                uid_to_non_zero_incentive.items(),
                key=lambda x: x[1],
                reverse=True,
            )[: self.hparams.max_topk_peers]

            # Extract just the UIDs (the first element of each tuple)
            top_incentive_peers = [int(uid) for uid, _ in top_incentive_peers]
```

To me this is a bit clearer what is happening (also using list instead of `np.array` if we decide to go back to that). Wdyt?

noaonoszko replied:
The lambda thing sure, but skeptical to using list instead of array, replied to that [here](https://github.com/tplr-ai/templar/pull/158#discussion_r2021419577).

noaonoszko replied:
Lambda nitpick fixed in 9445f69.

joellidin replied:
You would still need to filter out the uid from the tuple with this approach. It is more readable imo when you do it like this.

joellidin replied:
Ignore it and revert if you think it is a nitpick.

noaonoszko replied:
Damn, reverted in b0697dc.

joellidin commented:
One thing else. Because of the numpy type now we get this error:

```bash
Traceback (most recent call last):
  File "/home/bootstrap/templar/neurons/miner.py", line 786, in <module>
    asyncio.run(Miner().run())
  File "/home/bootstrap/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/home/bootstrap/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/bootstrap/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/bootstrap/templar/neurons/miner.py", line 699, in run
    "skipped_peers": json.dumps(
                     ^^^^^^^^^^^
  File "/home/bootstrap/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/bootstrap/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/bootstrap/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/home/bootstrap/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type int64 is not JSON serializable
```

Which crashes the miner.

noaonoszko commented:
> Which crashes the miner. Changing the peers to a ndarray propagates all over the code so I would say we should keep it as a list.
Are you running the latest code? I fixed that here:
```python
"gather_peers": json.dumps(self.comms.peers.tolist()),
```

noaonoszko commented:
Wdym "Changing the peers to a ndarray propagates all over the code so I would say we should keep it as a list."? If it's only about json dumping it without calling `tolist()` first I don't think it warrants delaying the PR further. Would require quite a lot of code changes.

noaonoszko commented:
Hm I see what you mean now:
```python
                        skipped_uids.append(uid)
```

It's not very nice to have different lists of uids of different types. I'll see if I can make them more consistent quickly.

joellidin commented:
> How about we just make it run now and then we make the types of the different collections of uids more consistent in a separate PR?

Sure, I can fix that in a separate PR. I just don't want to have edge cases which we don't catch for these tests. But then try to make sure all references from the uids from the gather result with a `tolist`. I think it is mainly the logging that would be broken and crash the miner.

joellidin commented:
Also if you could just do `ruff check --fix` for my OCD?

noaonoszko commented:
> Also if you could just do ruff check --fix for my OCD?
Done

noaonoszko commented:
> Sure, I can fix that in a separate PR. I just don't want to have edge cases which we don't catch for these tests. But then try to make sure all references from the uids from the gather result with a tolist. I think it is mainly the logging that would be broken and crash the miner.

The code ran before I updated the branch today so don't think there should be anything else. Also checked now and I couldn't find anything else from `gather_results` that needs to be addressed.

Also I plan to run the code again after rebase so that would catch anything. You're just running the aggregator right?

noaonoszko commented on src/tplr/neurons.py:
```diff
@@ -22,8 +22,19 @@
+
+sys.path.pop(0)  # Clean up immediately to avoid conflicts later
+#####
+
```
Nice.

noaonoszko replied:
Done in 1917883

joellidin commented:
Feel free to squash the commits whenever you want if there is not too much conflicts

joellidin commented:
Also, the validator crashes with.

```bash
Traceback (most recent call last):
  File "/home/bootstrap/templar/neurons/validator.py", line 1880, in <module>
    asyncio.run(Validator().run())
  File "/home/bootstrap/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/home/bootstrap/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/bootstrap/.local/share/uv/python/cpython-3.11.11-linux-x86_64-gnu/lib/python3.11/asyncio/base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/bootstrap/templar/neurons/validator.py", line 400, in run
    selected_peers = self.select_next_peers()
                     ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/bootstrap/templar/neurons/validator.py", line 1720, in select_next_peers
    active_non_zero_weight_uids = np.intersect1d(
                                  ^^^^^^^^^^^^^^^
  File "/home/bootstrap/templar/.venv/lib/python3.11/site-packages/numpy/lib/_arraysetops_impl.py", line 661, in intersect1d
    aux.sort()
TypeError: '<' not supported between instances of 'int' and 'set'
```
